### PR TITLE
Audio writers

### PR DESCRIPTION
**Pull request by EliteTK**
This pull request implements all the audio writers code. It is in its early stages and I think some improvements can be made but it will be easier to make those improvements if the code is upstreamed.

This also opens up the possibility for easily implementing lots of other output formats including ogg and wav.

### CODE REVIEW

File: audio.c [hunk 1]
```diff
@@ -88,98 +88,26 @@ writer(void *priv)
...
-		lame_set_asm_optimizations(lame, SSE, 1);
-		lame_set_quality(lame, 3);
-		lame_set_bWriteVbrTag(lame, 1);
-		lame_set_VBR(lame, vbr_mtrh);
-		lame_set_VBR_q(lame, 3);
-		lame_set_VBR_min_bitrate_kbps(lame, 96);
-		lame_set_VBR_max_bitrate_kbps(lame, 320);
+		aw = audio_writer_lame(ctx.fout, ctx.stream->sample_rate,
+					   ctx.stream->layout.channel_count,
+					   BUF_TIME_S, ctx.mono);
```
Comment (dhobsd) [inline audio.c]: I prefer using something stylistically similar to FreeBSD's style(9). There is some deviance in my interpretation (in particular, all control flow structures must use blocks, maximum line width is a little fluid around 100 characters, not 80). In this style, statements continue to the next line with four spaces. So for this code, the suggestion would be e.g.

```c
        aw = audio_writer_lame(ctx.fout, ctx.stream->sample_rate,
            ctx.stream->layout.channel_count, BUF_TIME_S, ctx.mono);
```

File: audio/writer-lame.c [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	assert(data != NULL);
+
+	lame = to_lame(writer);
+
+	left = data;
+	right = data + size / 2;
+	nsamples = size / bytes_per_frame;
+
+	switch (fmt) {
+	case SoundIoFormatFloat32LE: case SoundIoFormatFloat32BE:
```
Comment (dhobsd) [inline audio/writer-lame.c]: Another stylistic thing, please keep each case label on its own line here and below. I'm not a huge fan of the vertical size, but it is easier for me to read.

File: audio/writer-lame.c [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+};
+#define to_lame(ptr) container_of((ptr), struct lame, writer)
+
+static void
+lame_write(struct audio_writer *writer, enum SoundIoFormat fmt,
+	   char *data, int size, int bytes_per_frame)
+{
+	struct lame *lame;
+	char *left, *right;
+	int nsamples, blen = 0;
```
Comment (dhobsd) [inline audio/writer-lame.c]: Another stylistic thing, please stage declarations in order of line length from longest to shortest. In this example, `int nsamples...` and `struct lame ...` would be switched.

File: audio/writer-lame.c [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+	fwrite(lame->buf, 1, blen, lame->outfile);
+
+	lame_close(lame->lflags);
+	free(lame->buf);
+	free(lame);
+}
+
+struct audio_writer *
+audio_writer_lame(FILE *outfile, int sample_rate, int nchannels,
+		  int buf_time_s, int mono)
```
Comment (dhobsd) [inline audio/writer-lame.c]: Line continues at 4 spaces here, too.

File: audio/writer-lame.h [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#ifndef AUDIO_WRITER_LAME_H
+#define AUDIO_WRITER_LAME_H
+
+#include "writer.h"
+
+#ifdef WITH_LAME
+struct audio_writer *audio_writer_lame(FILE *outfile, int sample_rate, int nchannels, int buf_time_s, int mono);
+#else
+#define audio_writer_lame(...) (NULL)
```
Comment (dhobsd) [inline audio/writer-lame.h]: I think this illustrates that these should probably be modular and probably loaded with dlopen(3) either automatically or based on a configuration. I have some idea how I'd like that to look if you want to do that later, but this is surely good enough for now.

File: audio/writer.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#ifndef AUDIO_WRITER_H
+#define AUDIO_WRITER_H
+
+#include <stdio.h>
+#include <soundio/soundio.h>
+
+struct audio_writer {
+	void (*write)(struct audio_writer *writer, enum SoundIoFormat fmt,
+		      char *data, int size, int bytes_per_frame);
```
Comment (dhobsd) [inline audio/writer.h]: Should we `typedef` these function pointers? I'm not normally a fan of `typedef`, but I do find it makes function pointers easier to work with in all contexts.

File: audio.c [hunk 7]
```diff
@@ -88,98 +88,26 @@ writer(void *priv)
...
-
-				fwrite(mp3buf, 1, blen, ctx.fout);
-			} else {
-				size_t amt = fwrite(read_buf, 1, fill_bytes, ctx.fout);
-				if ((int)amt != fill_bytes) {
-					perror("fwrite");
-					exit(EXIT_FAILURE);
-				}
-			}
+			audio_writer_write(aw, ctx.stream->format, read_buf,
```
Comment (dhobsd) [inline audio.c]: Wrapping at 4 spaces here too, please.

File: audio.c [hunk 8]
```diff
@@ -88,98 +88,26 @@ writer(void *priv)
...
+					   fill_bytes,
+					   ctx.stream->bytes_per_frame);
```
Comment (dhobsd) [inline audio.c]: I believe with wrapping, these lines can be merged.

File: audio/writer.h [hunk 9]
```diff
@@ -0,0 +1,26 @@
+#ifndef AUDIO_WRITER_H
+#define AUDIO_WRITER_H
+
+#include <stdio.h>
+#include <soundio/soundio.h>
+
+struct audio_writer {
```
Comment (dhobsd) [inline audio/writer.h:7]: Elsewhere, you are embedding `struct audio_writer` into the per-writer structures. I would rather have `struct audio_writer` contain an opaque pointer to the context of the thing it is using. This does mean an additional allocation, but it gets rid of `container_of`, which I don't like very much. The other thing is that it makes it easier for us to transition `struct audio_writer` to something like a `struct audio_module`, which I think would be a nice way to move with this interface (as mentioned above).

**Review by dhobsd — CHANGES_REQUESTED** [review]
Functionally, this looks great. Only a few comments relating to style and a small functional change about the audio module. Thanks for doing this!

File: audio.c (hunk 8 above)
Comment (EliteTK) [reply audio.c]: If the limit is 100 characters, fill_bytes could be appended to line 108.

**Review by dhobsd — APPROVED** [review]

**Comment by dhobsd** [comment]
Thanks for this!

**Pull request merged**